### ISSUE TITLE
Unable to log in to Hypothesis from USA Today pages in Via

### ISSUE THREAD

**Issue by mattdricker** [hypothesis/viahtml#333]
**Update 2022-04-19: The problem with usatoday.com and the Cross-Origin-Opener-Policy header does not affect Via at present. It only affects the extension and bookmarklet. See https://github.com/hypothesis/viahtml/issues/333#issuecomment-1102480587 for explanation.**

### Steps to reproduce
1. Make sure you are logged out Hypothesis
2. Open any page at USA Today e.g. https://www.usatoday.com/story/news/politics/2022/03/07/ukraine-russia-invasion-live-updates/9404740002/
3. Activate Hypothesis extension or bookmarklet, and click the sidebar link to Log In

### Expected behaviour
The Hypothesis login pop-up window should appear with appropriate username and password fields available.

### Actual behaviour
An entirely blank pop-up window displays.

### Browser/system information
Tested on:

- Chrome 99
- Firefox 97
- Brave 1.36.109

Note: The issue does not seem to affect Safari 15.3

### Additional details
If you are already logged in to Hypothesis all functionality works as expected.

**Comment by robertknight** [comment]
BroadcastChannel is supported in Safari 15.4 but I encountered an issue with it being unable to be used for popup => opener communication, whereas in Chrome and Firefox this works. This might relate to origin partitioning that is implemented in Safari and being planned for Chrome. See https://github.com/whatwg/html/issues/5803.

**Comment by robertknight** [comment]
Some related discussion about the Cross-Origin-Opener-Policy specs and popups here: https://github.com/whatwg/html/issues/7713. From the linked [explainer](https://github.com/hemeryar/explainers/blob/main/coop_popups.md):

> The solution we propose is to add (yet) another COOP value: Popups (name likely to change). It says: "Pages that set a different COOP policy or have different top level origin will be placed in a different BrowsingContext group. However the opener relationship will not be severed, but the only windowProxy properties that remain accessible through it are postMessage and closed." Served together with a suitable COEP policy it would enable crossOriginIsolated.

That unfortunately might not help us since we can't control the COOP policy on the top-level document in all cases, though we can control it in Via.

**Comment by robertknight** [comment]
The situation with standards is fluid, but https://github.com/whatwg/html/issues/6364 has the most up-to-date discussion. My understanding of where this is likely to end up is:

1. Browsers will eventually provide a way for a popup window to communicate with a cross-origin frame that opens it via `window.opener`, even if that iframe or its parent document are using `Cross-Origin-Opener-Policy`. The `window.opener` reference would be limited to just use of the `postMessage` and `closed` properties, which would be adequate for us. Enabling this might require the parent document or popup to set a custom `Cross-Origin-Opener-Policy` headers. If headers have to be changed on the site where the client was embedded, as opposed to the URL opened in the popup window, that's more complicated and would depend on the method of embedding the client:
   - In Via we can override headers on any document.
   - In the bookmarklet we can't. This would add another category of site (in addition to those that use Content-Security-Policy) where the bookmarklet simply can't work.
   - In the Chrome extension we'd have to ask for additional permissions, which is undesirable, so I think we'd use some other extension-specific mechanism to handle the login
2. In today's browsers, `BroadcastChannel` can be used as a workaround for pages that use COOP in Firefox and Chrome but not Safari. The same applies to other mechanisms of same-origin communication such as `localStorage`. `BroadcastChannel` is liable to stop working in future though as Firefox/Chrome are planning to partition it similar to Safari. Hopefully by then they will have implemented (1) however
3. It might be possible to use `document.requestStorageAccess()` together with `document.cookie` to allow the popup to communicate with the iframe in Safari/Firefox. I need to prototype this.

So the status quo seems to be that there is no solution that works in all browsers, all methods of delivering the client and does not affect the page's ability to use web platform features that are depend on COOP.

Some options solutions we can implement today:

 1. Strip the `Cross-Origin-Opener-Policy` header in Via, and rely on sites not actively using the web platform features that are gated behind it. When the solution described in (1) is available, we could change the header to one that allows popups rather than strip it entirely.
 2. Use `BroadcastChannel` or `document.cookie` as popup <-> iframe communication mechanisms, with the understanding that these may become unavailable in future.
 3. Use an alternative method of communicating the access token from the popup to iframes in the extension. This will involve extension-specific APIs.

**Comment by chrisshaw** [comment]
@robertknight Focusing only on Via for this ticket (Extension covered in https://github.com/hypothesis/product-backlog/issues/1337):

- @mattdricker do you still see this issue on USA Today? Am I crazy, or did they change their COOP? (I also am unable to reproduce [the archlinux.org example](https://github.com/hypothesis/viahtml/issues/334))
  ![image](https://user-images.githubusercontent.com/2966918/163744138-2c9dc73b-cab9-4ae2-a771-7b5b5f7d0604.png)
- @robertknight Let's assume for a moment this is a problem on other sites: if we have them complete a workaround where they go log in on a site that does not use this COOP (maybe even one we host), can they then come back to this Via page, would that work to "log them in?" I'm starting to think the best answer here is likely to point people to a workaround and let the standards solidify, rather than stripping out headers that sites likely don't want stripped out. Especially if neither report is no longer an issue.

**Comment by robertknight** [comment]
In general Via does not strip the `Cross-Origin-Opener-Policy` header. If test a local instance of Via with the client's COOP test page at http://localhost:3000/document/coop-test then I see that viahtml preserves the `Cross-Origin-Opener-Policy: same-origin` header. Searching through pywb's source I didn't find any relevant mentions of COOP either.

With the original test URL (https://www.usatoday.com/story/news/politics/2022/03/07/ukraine-russia-invasion-live-updates/9404740002/) however I do see a difference. To avoid the redirect mentioned in https://github.com/hypothesis/viahtml/issues/333#issuecomment-1096905425 I ran these tests from one of our AWS servers:

```
curl -i 'https://www.usatoday.com/story/news/politics/2022/03/07/ukraine-russia-invasion-live-updates/9404740002/' 
```

Returns a response with both `cross-origin-resource-policy: same-origin` and `cross-origin-opener-policy: same-origin` headers

```
curl -H 'Referer: https://via.hypothes.is' -i 'https://viahtml.hypothes.is/https://www.usatoday.com/story/news/politics/2022/03/07/ukraine-russia-invasion-live-updates/9404740002/'
```

Returns a response with a `cross-origin-resource-policy: same-origin` header but no `cross-origin-opener-policy: same-origin` header.

**Comment by robertknight** [comment]
It turns out that one of the [engineers](https://arturjanc.com) working on these specifications has created a [test page](https://arturjanc.com/cgi-bin/coop/index-with-coop.py?coop=same-origin) which can be used for testing this with Via. See https://via.hypothes.is/https://arturjanc.com/cgi-bin/coop/index-with-coop.py?coop=same-origin

Inspecting the requests made when accessing this page through Via, I can see that Via has preserved the `cross-origin-opener-policy: same-origin` header from the original site. So Via is not, in general, stripping this header. However, the user can successfully login! How comes? I believe this is because of the way Via is structured: `Cross-Origin-Opener-Policy` only takes effect if set on the top-level frame, but in the public Via the HTML content is currently always displayed inside a child frame. As a result the COOP header is ignored and login still works. I tested this in Safari and Firefox with the same results. In the LMS's Via the structure is different, Via redirects to ViaHTML rather than embedding it in an iframe. In this context the COOP header from the top-level frame, which is the LMS, would apply. That is outside of our control so perhaps this could become an issue in future.

The lack of COOP header on the top-level frame may eventually cause problems for some sites if they rely on using features which are only available when these headers are set. Today however this is not an issue.

If instead of using Via you visit https://arturjanc.com/cgi-bin/coop/index-with-coop.py?coop=same-origin directly and activate the extension, you'll find that login fails. What you can do for the extension is visit some other website, login there and then reload the page where COOP is set. This will continue to work in Chrome until third-party cookies are phased out. In Safari however third-party cookies are blocked and so this doesn't work.

So to summarize: contrary to what I wrote earlier, it looks like the Cross-Origin-Opener-Policy header is not a problem for public Via today. This is because the header must be set on the top-level frame to take effect. In Via however, the proxied headers from the upstream site apply to the inner iframe where ViaHTML document is being loaded. In the LMS's Via we might eventually run into the problem again if the LMS itself enables the COOP header. I don't expect that will happen soon.

**Comment by robertknight** [comment]
We've [agreed to close this issue](https://hypothes-is.slack.com/archives/C2C2U40LW/p1650365629637269) for now since the problem doesn't currently affect Via. For the Chrome extension there is a workaround which is to log in using a different site. I will keep an eye on the relevant standards discussions and create / update issues as needed.
